text/0069-api-v2.md

```diff
@@ -0,0 +1,178 @@
+
+Keys in other keyspace will be invisible to the user.
+
+## Reference-level design
+
+### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes and only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+### RawKV Value Encode
+
+If the key is started with `r` or `a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### PD
+
+Add a new http interface for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    prefix_id: "0"
+  }, 
+  {
+    name: "graghdb",
+    // a prefix id can be referred from mulitple keyspace name
+    prefix_id: "0"    
+  }, 
+  {
+    name: "redis",
+    prefix_id: "1"
+  }
+]
+
+// add new keyspace
+POST /keyspaces
+{
+  name: "mongodb",
+  // optional. if is null, PD will allocate an unused prefix id
+  prefix_id: "2"
+}
+
+// update a keyspace
+POST /keyspaces/{keyspace_name}
+// delete a keyspace
+DELETE /keyspaces/{keyspace_name}
+```
+
+The keyspace is stored in etcd and should have no limitation on the numbers.
+
+### TiKV
+
+In TiKV config file, add a new configuration `storage.api_version`. When enabled, `storage.enable_ttl` must also be enabled.
+
+In kvdb, add a store meta `api_version`. When the store meta mismatches the config`storage.enable_ttl`, it means that the user is switching the API version, then check no non-TiDB exist, and then save the new api version in store meta.
+
+In kvproto message `SSTMeta`, add `api_version`. Reject the SST file if the version is mismatched.
+
+In TiKV gRPC's context, add a field `api_version`.
+
+If `storage.api_version=2`:
+
+- Run TTL compaction filter only on the keys starting with `r`.
+
+- Use the `API V2` Value encode in `RawStore`, `TTLStore` and `sst_importer`.
+
+- If the request's context has `api_version=2`:
+  - Reject the request except for it's a TxnKV request and the keys starting with `m` or `t`.
```

Comment:
why rejecting these requests since the `storage.api_vesion` and the request's context has `api_version=2`?

Reply:
Fixed. It was a typo, I was meant to say "If the request's context has `api_version=1`".